File: client/web/src/features/devices/api/devices-api.ts
import { apiClient } from "@/lib/api-client";
import type { DeviceInfo, Envelope } from "@/types/api-types";

export interface DeviceListResponse {
  devices: DeviceInfo[];
}

export function fetchDevices(): Promise<DeviceListResponse> {
  return apiClient.get<DeviceListResponse>("/devices");
}

export function approveDevice(
  deviceId: string,
  approvalRequestId: string,
  deviceEnvelope: Envelope,
): Promise<void> {
  return apiClient.post<void>(`/devices/${deviceId}/approve`, {
    approvalRequestId,
    approvalAuth: { type: "device" },
    deviceEnvelope,
  });
}

export function rejectDevice(deviceId: string): Promise<void> {
  return apiClient.post<void>(`/devices/${deviceId}/reject`);
}

export function revokeDevice(deviceId: string): Promise<void> {
  return apiClient.post<void>(`/devices/${deviceId}/revoke`);
}

export function updateSelfEnvelope(envelope: Envelope): Promise<void> {
  return apiClient.put<void>("/devices/self/envelope", envelope);
}

File: client/android/core/src/main/java/io/github/pstanar/pstotp/core/api/DevicesApi.kt
package io.github.pstanar.pstotp.core.api

import io.github.pstanar.pstotp.core.model.api.DeviceListResponse
import io.github.pstanar.pstotp.core.model.api.Envelope
import org.json.JSONObject

/** Device management API endpoints (authenticated). */
class DevicesApi(private val client: ApiClient) {

    suspend fun fetchDevices(): DeviceListResponse =
        client.get("/devices") { DeviceListResponse.fromJson(it) }

    suspend fun approveDevice(
        deviceId: String,
        approvalRequestId: String,
        deviceEnvelope: Envelope,
    ) {
        val body = JSONObject().apply {
            put("approvalRequestId", approvalRequestId)
            put("approvalAuth", JSONObject().put("type", "device"))
            put("deviceEnvelope", deviceEnvelope.toJson())
        }
        client.post("/devices/$deviceId/approve", body) { }
    }

    suspend fun rejectDevice(deviceId: String) {
        client.post("/devices/$deviceId/reject", null) { }
    }

    suspend fun revokeDevice(deviceId: String) {
        client.post("/devices/$deviceId/revoke", null) { }
    }

    suspend fun updateSelfEnvelope(envelope: Envelope) {
        client.put("/devices/self/envelope", envelope.toJson()) { }
    }
}
